### ISSUE TITLE
Inaccessible markdown heading links

### ISSUE THREAD

**Issue by zmrhaljiri** [facebook/docusaurus#8336]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [ ] I have tried the `npm run clear` or `yarn clear` command.
- [ ] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [ ] I have tried creating a repro with https://new.docusaurus.io.
- [ ] I have read the console error message carefully (if applicable).

### Description

Heading hash links in Markdown docs are missing the [accessible name](https://www.w3.org/WAI/ARIA/apg/practices/names-and-descriptions#introduction). Screen reader users who tab through the page by keyboard and focus on the links hear only "Direct link to heading" (the text inside the `title` attribute of hash links).

Also, when users of screen readers populate a list of links on the page, it is showing the following:

![nvda-link-list-docusaurus](https://user-images.githubusercontent.com/19836714/201651387-260bf083-4b4b-4ee2-b575-81a67f9ad0b7.jpg "Elements list window of the NVDA screen reader showing list of all links on a page. The text content of the heading hash links is just the '#' character, instead of the actual heading text.")

Basically, they do not know where the links will take them since there is no meaningful text content.

### Reproducible demo

https://docusaurus.io/docs#fast-track

### Steps to reproduce

1. Go to any docusaurus docs page, for example, https://docusaurus.io/docs
2. Turn on the screen reader (e.g. NVDA on Windows, or VoiceOver on Mac)
3. Tab through the heading links:
  - Fast Track (the first hash link)
  - Docusaurus: Documentation Made Easy (the second hash link)
  - and so on
4. When focused on the link, the screen reader announces something like "Direct link to heading, link"

### Expected behavior

The hash links should have an accessible name, ideally the heading content, so screen reader users know where the links are pointing to.

The suggestion for the new HTML structure of a sample link:

```html
<h2 id="fast-track">
  <a href="#fast-track" title="Direct link to Fast track">Fast track</a>
</h2>
```

The implementation of the `#` character can be still made on hover, similarly as on [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#try_it).

### Actual behavior

The hash links have no accessible name (no unique text content) so screen reader users do not know where the links are pointing to.

The current HTML structure of a sample link:

```html
<h2 id="fast-track">
  Fast Track 
  <a class="hash-link" href="#fast-track" title="Direct link to heading">&ZeroWidthSpace;</a>
</h2>
```

### Your environment

- Public site URL: https://docusaurus.io/docs#fast-track
- Docusaurus version used: 2.2.0

### Self-service

- [ ] I'd be willing to fix this bug myself.

**Comment by slorber** [comment]
Makes sense, thanks for reporting.


> The hash links should have an accessible name, ideally the heading content,

> The implementation of the # character can be still made on hover, similarly as on [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#try_it).

Is this mandatory or an official a11y reco? 

Many other solutions are not adding the link on the full heading, see for example:
- https://squidfunk.github.io/mkdocs-material/getting-started/
- https://vitepress.vuejs.org/guide/what-is-vitepress

Personally, I don't find the MDN implementation pleasant to look at for a regular user without any disability, and would prefer to keep the existing pattern and just fix the title, is it fine?

---


Note we already have a translated anchor title:

```tsx
<a
        className="hash-link"
        href={`#${id}`}
        title={translate({
          id: 'theme.common.headingLinkTitle',
          message: 'Direct link to heading',
          description: 'Title for link to heading',
        })}>
        &#8203;
</a>
```

Do you have a recommendation for the label so that we do not have to "rewrite" all the existing translations from scratch? (I personally can only translate EN + FR)

Is something like "Direct link to heading - {headingLabel}" good enough?

**Comment by zmrhaljiri** [comment]
Thank you for your reply!

IMHO it would be a violation of the [WCAG success criterion 2.4.4. Link Purpose (In Context) (A)](https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html), however, the user experience for users of assistive technologies is currently as described, so any approach that results in links having an accessible and meaningful name would be fine. There is a good article about this problematic: https://amberwilson.co.uk/blog/are-your-anchor-links-accessible/.

The text you're suggesting for the title is great. The main problem is we cannot use the `title` attribute as the only means for setting the accessible name of the link as it is infamously inaccessible (not supported by many screen readers). You can supply the helper text "Direct link to heading - {headingLabel}" into the `title`, but the link should have the text content (instead of `&#8203;`) in the first place.

**Comment by slorber** [comment]
What I see in practice is that tools like VoiceOver will announce better if the link has aria-label rather than using title. Isn't this a good-enough solution?

Duplicating the link title in a hidden element feels like a weird solution to me 🤷‍♂️ 

![CleanShot 2022-11-23 at 18 47 38@2x](https://user-images.githubusercontent.com/749374/203614945-59a9ad0d-7ed0-4d8a-8d01-9630558cbaf0.png)

Will read that carefully when I have time but if anyone can suggest a good-enough solution that does not involve useless hidden child elements duplication + making the whole text the anchor link, I'd be happy to implement it fast.

**Comment by slorber** [comment]
Can you tell exactly how we violate the spec according to you? 

Unfortunately I don't have much time to read in details the whole docs 😅

**Comment by slorber** [comment]
Thanks for the explanations!

> Another approach can be the aria-label way you suggested, which would also work since the accessible name of the link would not be empty space or "#" character, but "Direct link to Fast track":

I guess we can move on and just add this new aria-label then, that looks like a good enough solution

> Sorry, I missed your points about the other solutions before

Not saying those linked examples are good accessibility examples, just wanted to illustrate other docs sites preferring not having clickable anchor headings.

> Thank you for taking the time looking to this, really. I realize it can be quite tricky and complex. I am currently building a site dedicated to web accessibility using Docusaurus, and I'd greatly appreciate if accessibility issues and difficulties can be handled

We also appreciate having a base of users caring about accessibility, as this helps up improve on topics we may not have expertise on 😄 

> optionally have control over the markup so I can adjust it.

In any case, you should be able to swizzle headings components and implement your own markup, but we should have good accessibility by default. 

If aria-label is not good enough for you, you can swizzle and make it use heading anchor links instead.

**Comment by zmrhaljiri** [comment]
Thanks a lot, much appreciated! If `aria-label` will be there, no need for me to swizzle then :) 💚
